pmlc/conversion/comp_to_llvm/comp_to_vulkan.cc

```diff
@@ -0,0 +1,472 @@
+  }
+};
+
+struct ConvertScheduleFunc : ConvertCompToVkBasePattern<comp::ScheduleFunc> {
+  ConvertScheduleFunc(const BinaryModulesMap &modulesMap,
+                      mlir::TypeConverter &typeConverter,
+                      mlir::MLIRContext *context)
+      : ConvertCompToVkBasePattern<comp::ScheduleFunc>(typeConverter, context),
+        modulesMap(modulesMap) {}
+
+  mlir::LogicalResult
+  matchAndRewrite(comp::ScheduleFunc op, mlir::ArrayRef<mlir::Value> operands,
+                  mlir::ConversionPatternRewriter &rewriter) const override;
+
+  const BinaryModulesMap &modulesMap;
+};
+
+} // namespace
+
+void populateCompToVkPatterns(mlir::MLIRContext *context,
+                              const BinaryModulesMap &modulesMap,
+                              mlir::TypeConverter &typeConverter,
+                              mlir::OwningRewritePatternList &patterns) {
+  // Populate type conversion patterns.
+  LLVM::LLVMType llvmInt8Ptr = LLVM::LLVMType::getInt8PtrTy(context);
+  typeConverter.addConversion(
+      [=](comp::ExecEnvType execEnvType) -> mlir::Optional<mlir::Type> {
+        return llvmInt8Ptr;
+      });
+  typeConverter.addConversion(
+      [=](comp::EventType eventType) -> mlir::Optional<mlir::Type> {
+        return llvmInt8Ptr;
+      });
+  patterns.insert<ConvertToInitVulkan>(kInitVulkan, typeConverter, context);
+  patterns.insert<ConvertToDeinitVulkan>(kDeinitVulkan, typeConverter, context);
+  patterns.insert<ConvertToCommandBufferSubmit>(kRun, typeConverter, context);
+
+  patterns.insert<ConvertScheduleFunc>(modulesMap, typeConverter, context);
+  patterns.insert<ConvertAlloc>(typeConverter, context);
+
+  // TODO modify those to vulkan api
+  patterns.insert<ConvertDealloc>(kOclDealloc, typeConverter, context);
+  patterns.insert<ConvertScheduleBarrier>(kOclBarrier, typeConverter, context,
+      /*varArg=*/true, /*nonVarArgs=*/1);
+  patterns.insert<ConvertWait>(kOclWait, typeConverter, context,
+      /*varArg=*/true, /*nonVarArgs=*/0);
+
+  patterns.insert<ConvertScheduleRead>(kOclRead, typeConverter, context);
+  patterns.insert<ConvertScheduleWrite>(kOclWrite, typeConverter, context);
+
+}
+
+void addVkFunctionDeclarations(mlir::ModuleOp &module) {
+  mlir::Location loc = module.getLoc();
+  mlir::OpBuilder builder(module.getBody()->getTerminator());
+  mlir::MLIRContext *context = builder.getContext();
+  LLVM::LLVMType llvmInt8Ptr = LLVM::LLVMType::getInt8PtrTy(context);
+  LLVM::LLVMType llvmVoid = LLVM::LLVMType::getVoidTy(context);
+  LLVM::LLVMType llvmInt32 = LLVM::LLVMType::getInt32Ty(context);
+  LLVM::LLVMType llvmInt64Type = LLVM::LLVMType::getInt64Ty(context);
+  mlir::Type mlirIndexType = builder.getIndexType();
+
+  builder.create<LLVM::LLVMFuncOp>(
+      loc, kInitVulkan,
+      LLVM::LLVMType::getFunctionTy(llvmInt8Ptr, {llvmInt8Ptr},
+          /*isVarArg=*/false));
+
+  builder.create<mlir::FuncOp>(
+      loc, kCreateVulkanLaunchKernelAction,
+      mlir::FunctionType::get(
```

Comment:
I think it is better to use llvmInt32 Type than mlir type, so we can use LLVM::LLVMFuncOp insteadof mlir::FuncOp.
Please see
https://github.com/plaidml/plaidml/pull/1356/files#diff-b5c69b0848345bd7a7b84e4e3d744e49R490
https://github.com/plaidml/plaidml/pull/1356/files#diff-b5c69b0848345bd7a7b84e4e3d744e49R588
https://github.com/plaidml/plaidml/pull/1356/files#diff-bd02d185b766dd5f7de33b0148b25089L116